File: src/sweep_dreams/domain/calendar.py
import calendar
from datetime import datetime, timedelta
from zoneinfo import ZoneInfo

from sweep_dreams.domain.models import (
    PACIFIC_TZ,
    ParkingRegulation,
    RecurringRule,
    BlockSchedule,
    SweepingSchedule,
    Weekday,
    _WEEKDAY_LOOKUP,
)


def _nth_weekday(year: int, month: int, weekday: int, occurrence: int) -> int | None:
    """Return the day of the month for the nth occurrence of a weekday."""
    first_weekday, days_in_month = calendar.monthrange(year, month)
    day = 1 + ((weekday - first_weekday) % 7) + 7 * (occurrence - 1)
    return day if day <= days_in_month else None


def _normalize_now(now: datetime | None, tz: ZoneInfo) -> datetime:
    if now is None:
        return datetime.now(tz)
    if now.tzinfo is None:
        return now.replace(tzinfo=tz)
    return now.astimezone(tz)


def next_sweep_window_from_rule(
    rule: RecurringRule,
    *,
    now: datetime | None = None,
    tz: ZoneInfo | None = None,
) -> tuple[datetime, datetime]:
    """
    Compute the next sweeping window (start, end) for a RecurringRule.

    Core scheduling logic extracted for reuse with BlockSchedule.
    """
    tzinfo = tz or PACIFIC_TZ
    reference = _normalize_now(now, tzinfo)

    # Extract weekday from pattern (should only have one)
    if not rule.pattern.weekdays:
        raise ValueError("Rule has no weekdays configured.")
    weekday = next(iter(rule.pattern.weekdays))

    # Get active weeks (None means all weeks 1-5)
    active_weeks = (
        sorted(rule.pattern.weeks_of_month)
        if rule.pattern.weeks_of_month
        else [1, 2, 3, 4, 5]
    )
    if not active_weeks:
        raise ValueError("Rule has no active weeks configured.")

    start_hour = rule.time_window.start.hour
    end_hour = rule.time_window.end.hour

    for month_offset in range(0, 13):
        month_index = reference.month - 1 + month_offset
        year = reference.year + month_index // 12
        month = (month_index % 12) + 1

        for occurrence in active_weeks:
            day = _nth_weekday(year, month, weekday, occurrence)
            if day is None:
                continue

            start_dt = datetime(year, month, day, start_hour, tzinfo=tzinfo)
            end_dt = datetime(year, month, day, end_hour, tzinfo=tzinfo)
            if end_dt <= start_dt:
                end_dt += timedelta(days=1)  # Handle windows that cross midnight.

            if end_dt <= reference:
                continue  # Window already passed.

            if start_dt <= reference <= end_dt or start_dt > reference:
                return start_dt, end_dt

    raise ValueError("Unable to compute next sweep window within 12 months.")


def next_sweep_window(
    schedule: SweepingSchedule,
    *,
    now: datetime | None = None,
    tz: ZoneInfo | None = None,
) -> tuple[datetime, datetime]:
    """
    Compute the next sweeping window (start, end) for a schedule.
    """
    tzinfo = tz or PACIFIC_TZ
    reference = _normalize_now(now, tzinfo)

    weekday_label = (schedule.week_day or "").strip().lower()
    if weekday_label == "holiday":
        raise ValueError(
            "Schedule applies only on holidays; next sweeping day is not defined."
        )
    if weekday_label not in _WEEKDAY_LOOKUP:
        raise ValueError(f"Unknown weekday label: {schedule.week_day!r}")
    weekday = _WEEKDAY_LOOKUP[weekday_label]

    active_weeks = [
        idx
        for idx, active in enumerate(
            [
                schedule.week1,
                schedule.week2,
                schedule.week3,
                schedule.week4,
                schedule.week5,
            ],
            start=1,
        )
        if bool(active)
    ]
    if not active_weeks:
        raise ValueError("Schedule has no active weeks configured.")
    if schedule.from_hour is None or schedule.to_hour is None:
        raise ValueError("Schedule is missing from/to hours.")

    for month_offset in range(0, 13):
        month_index = reference.month - 1 + month_offset
        year = reference.year + month_index // 12
        month = (month_index % 12) + 1

        for occurrence in sorted(active_weeks):
            day = _nth_weekday(year, month, weekday, occurrence)
            if day is None:
                continue

            start_dt = datetime(year, month, day, schedule.from_hour, tzinfo=tzinfo)
            end_dt = datetime(year, month, day, schedule.to_hour, tzinfo=tzinfo)
            if end_dt <= start_dt:
                end_dt += timedelta(days=1)  # Handle windows that cross midnight.

            if end_dt <= reference:
                continue  # Window already passed.

            if start_dt <= reference <= end_dt or start_dt > reference:
                return start_dt, end_dt

    raise ValueError("Unable to compute next sweep window within 12 months.")


def earliest_sweep_window(
    block_schedule: BlockSchedule,
    *,
    now: datetime | None = None,
    tz: ZoneInfo | None = None,
) -> tuple[datetime, datetime]:
    """
    Compute the earliest sweep window across all rules in a BlockSchedule.

    Args:
        block_schedule: The block schedule with multiple rules
        now: Reference time (defaults to current time)
        tz: Timezone (defaults to PACIFIC_TZ)

    Returns:
        Tuple of (start, end) datetime for the earliest sweep window

    Raises:
        ValueError: If no valid sweep windows can be computed
    """
    earliest_start = None
    earliest_end = None

    for rule in block_schedule.rules:
        try:
            start, end = next_sweep_window_from_rule(rule=rule, now=now, tz=tz)
        except ValueError:
            # Skip rules that can't compute windows (e.g., holiday-only)
            continue

        if earliest_start is None or start < earliest_start:
            earliest_start = start
            earliest_end = end

    if earliest_start is None:
        raise ValueError(
            f"Could not compute sweep window for block {block_schedule.block}"
        )

    return earliest_start, earliest_end


def earliest_sweep_window_with_block_id(
    block_schedule: BlockSchedule,
    source_schedules: list[SweepingSchedule],
    *,
    now: datetime | None = None,
    tz: ZoneInfo | None = None,
) -> tuple[datetime, datetime, int]:
    """
    Compute the earliest sweep window and return its block_sweep_id.

    Args:
        block_schedule: The merged block schedule
        source_schedules: The underlying SweepingSchedule rows for this block
        now: Reference time (defaults to current time)
        tz: Timezone (defaults to PACIFIC_TZ)

    Returns:
        (start, end, block_sweep_id) for the earliest window

    Raises:
        ValueError: If no valid sweep window can be computed
    """
    tzinfo = tz or PACIFIC_TZ
    reference = _normalize_now(now, tzinfo)

    earliest_start: datetime | None = None
    earliest_end: datetime | None = None
    earliest_block_sweep_id: int | None = None

    for schedule in source_schedules:
        try:
            start, end = next_sweep_window(schedule, now=reference, tz=tzinfo)
        except ValueError:
            continue

        is_better_start = earliest_start is None or start < earliest_start
        is_same_start = earliest_start is not None and start == earliest_start
        if is_better_start or (
            is_same_start
            and earliest_block_sweep_id is not None
            and schedule.block_sweep_id < earliest_block_sweep_id
        ):
            earliest_start = start
            earliest_end = end
            earliest_block_sweep_id = schedule.block_sweep_id

    if (
        earliest_start is None
        or earliest_end is None
        or earliest_block_sweep_id is None
    ):
        raise ValueError(
            f"Could not compute sweep window for block {block_schedule.block}"
        )

    return earliest_start, earliest_end, earliest_block_sweep_id


# Mapping for parking regulation day ranges
_DAYS_TO_WEEKDAYS: dict[str, set[Weekday]] = {
    "m-f": {Weekday.MON, Weekday.TUE, Weekday.WED, Weekday.THU, Weekday.FRI},
    "m-sa": {Weekday.MON, Weekday.TUE, Weekday.WED, Weekday.THU, Weekday.FRI, Weekday.SAT},
    "m-su": {Weekday.MON, Weekday.TUE, Weekday.WED, Weekday.THU, Weekday.FRI, Weekday.SAT, Weekday.SUN},
    "sa-su": {Weekday.SAT, Weekday.SUN},
    "su": {Weekday.SUN},
    "sa": {Weekday.SAT},
}


def _parse_military_time(military: int) -> tuple[int, int]:
    """Parse military time (e.g., 800 -> (8, 0), 1800 -> (18, 0))."""
    hours = military // 100
    minutes = military % 100
    return hours, minutes


def next_parking_regulation_window(
    regulation: ParkingRegulation,
    *,
    now: datetime | None = None,
    tz: ZoneInfo | None = None,
) -> tuple[datetime, datetime]:
    """
    Compute the next time window when a parking regulation applies.

    For time-limited parking, this is when the restriction is active
    (e.g., 8am-6pm M-F means you can't park longer than the hour limit).

    Args:
        regulation: The parking regulation
        now: Reference time (defaults to current time)
        tz: Timezone (defaults to PACIFIC_TZ)

    Returns:
        Tuple of (start, end) datetime for the next regulation window

    Raises:
        ValueError: If regulation has no valid time/day configuration
    """
    tzinfo = tz or PACIFIC_TZ
    reference = _normalize_now(now, tzinfo)

    # Parse the days field
    days_str = (regulation.days or "").strip().lower()
    if not days_str or days_str not in _DAYS_TO_WEEKDAYS:
        raise ValueError(f"Unknown or missing days pattern: {regulation.days!r}")
    active_weekdays = _DAYS_TO_WEEKDAYS[days_str]

    # Parse time range
    if regulation.hrs_begin is None or regulation.hrs_end is None:
        raise ValueError("Regulation is missing hrs_begin/hrs_end")

    start_hour, start_min = _parse_military_time(regulation.hrs_begin)
    end_hour, end_min = _parse_military_time(regulation.hrs_end)

    # Search up to 8 days ahead (covers all weekdays)
    for day_offset in range(8):
        candidate_date = reference.date() + timedelta(days=day_offset)
        candidate_weekday = Weekday(candidate_date.weekday())

        if candidate_weekday not in active_weekdays:
            continue

        start_dt = datetime(
            candidate_date.year,
            candidate_date.month,
            candidate_date.day,
            start_hour,
            start_min,
            tzinfo=tzinfo,
        )
        end_dt = datetime(
            candidate_date.year,
            candidate_date.month,
            candidate_date.day,
            end_hour,
            end_min,
            tzinfo=tzinfo,
        )

        # Handle windows that cross midnight
        if end_dt <= start_dt:
            end_dt += timedelta(days=1)

        # Skip if window has already passed
        if end_dt <= reference:
            continue

        # Found a valid window
        return start_dt, end_dt

    raise ValueError("Unable to compute next regulation window within 8 days.")

File: workers/src/lib/calendar.ts
import type { RecurringRule, SweepingSchedule } from '../models';
import { WEEKDAY_LOOKUP, Weekday } from '../models';

const PACIFIC_TZ = 'America/Los_Angeles';

/**
 * Parse a days string like "M-F", "M-Sa", "M-Su" into a set of weekdays.
 * Case-insensitive.
 *
 * @param days - Days string (e.g., "M-F", "M-Sa", "M-Su")
 * @returns Set of Weekday enum values
 */
export function parseDaysString(days: string): Set<Weekday> {
  const normalized = days.trim().toLowerCase();

  // Handle common patterns
  if (normalized === 'm-f') {
    return new Set([Weekday.MON, Weekday.TUE, Weekday.WED, Weekday.THU, Weekday.FRI]);
  }
  if (normalized === 'm-sa') {
    return new Set([Weekday.MON, Weekday.TUE, Weekday.WED, Weekday.THU, Weekday.FRI, Weekday.SAT]);
  }
  if (normalized === 'm-su') {
    return new Set([Weekday.MON, Weekday.TUE, Weekday.WED, Weekday.THU, Weekday.FRI, Weekday.SAT, Weekday.SUN]);
  }

  throw new Error(`Unknown days pattern: ${days}`);
}

/**
 * Find the day of the month for the nth occurrence of a weekday.
 * Example: nthWeekday(2025, 1, Weekday.MON, 2) returns the 2nd Monday in Jan 2025
 *
 * @param year - The year
 * @param month - The month (1-12)
 * @param weekday - The weekday enum value (0=Monday, 6=Sunday)
 * @param occurrence - Which occurrence (1-5)
 * @returns Day of month, or null if it doesn't exist
 */
function nthWeekday(year: number, month: number, weekday: Weekday, occurrence: number): number | null {
  // Get the first day of the month
  const firstDay = new Date(year, month - 1, 1);
  const firstWeekday = firstDay.getDay();  // 0 = Sunday in JavaScript

  // Convert JavaScript weekday (0=Sunday) to our enum (0=Monday)
  // JavaScript: Sun=0, Mon=1, Tue=2, ..., Sat=6
  // Our enum:   Mon=0, Tue=1, Wed=2, ..., Sun=6
  // Formula: (jsWeekday + 6) % 7 = ourWeekday
  const adjustedFirstWeekday = (firstWeekday + 6) % 7;

  // Calculate the day for the nth occurrence
  let day = 1 + ((weekday - adjustedFirstWeekday + 7) % 7) + 7 * (occurrence - 1);

  // Check if this day exists in the month
  const daysInMonth = new Date(year, month, 0).getDate();
  return day <= daysInMonth ? day : null;
}

/**
 * Compute the next sweep window (start, end) for a RecurringRule.
 * Port of Python's next_sweep_window_from_rule().
 *
 * @param rule - The recurring rule
 * @param now - Reference time (defaults to current time)
 * @returns Tuple of [start, end] datetimes
 */
export function nextSweepWindowFromRule(
  rule: RecurringRule,
  now?: Date,
): [Date, Date] {
  const reference = now || new Date();

  // Extract weekday from pattern (should only have one)
  if (rule.pattern.weekdays.length === 0) {
    throw new Error('Rule has no weekdays configured.');
  }
  const weekday = rule.pattern.weekdays[0];

  // Get active weeks (null means all weeks 1-5)
  const activeWeeks = rule.pattern.weeks_of_month
    ? [...rule.pattern.weeks_of_month].sort((a, b) => a - b)
    : [1, 2, 3, 4, 5];

  if (activeWeeks.length === 0) {
    throw new Error('Rule has no active weeks configured.');
  }

  const [startHour, startMin] = rule.time_window.start.split(':').map(Number);
  const [endHour, endMin] = rule.time_window.end.split(':').map(Number);

  // Search up to 13 months ahead
  for (let monthOffset = 0; monthOffset < 13; monthOffset++) {
    const monthIndex = reference.getMonth() + monthOffset;
    const year = reference.getFullYear() + Math.floor(monthIndex / 12);
    const month = (monthIndex % 12) + 1;

    for (const occurrence of activeWeeks) {
      const day = nthWeekday(year, month, weekday, occurrence);
      if (day === null) continue;

      // Create start/end datetimes
      // Note: JavaScript Date constructor uses 0-indexed months
      let startDt = new Date(year, month - 1, day, startHour, startMin || 0);
      let endDt = new Date(year, month - 1, day, endHour, endMin || 0);

      // Handle windows that cross midnight
      if (endDt <= startDt) {
        endDt = new Date(endDt.getTime() + 24 * 60 * 60 * 1000);
      }

      // Skip windows that already passed
      if (endDt <= reference) continue;

      // Return the first valid window (in progress or future)
      if (startDt <= reference || startDt > reference) {
        return [startDt, endDt];
      }
    }
  }

  throw new Error('Unable to compute next sweep window within 12 months.');
}

/**
 * Compute the next sweep window for a raw SweepingSchedule.
 * Port of Python's next_sweep_window().
 *
 * @param schedule - The sweeping schedule
 * @param now - Reference time (defaults to current time)
 * @returns Tuple of [start, end] datetimes
 */
export function nextSweepWindow(
  schedule: SweepingSchedule,
  now?: Date,
): [Date, Date] {
  const reference = now || new Date();

  const weekdayLabel = (schedule.week_day || '').trim().toLowerCase();
  if (weekdayLabel === 'holiday') {
    throw new Error('Schedule applies only on holidays; next sweeping day is not defined.');
  }

  const weekday = WEEKDAY_LOOKUP[weekdayLabel];
  if (weekday === undefined) {
    throw new Error(`Unknown weekday label: ${schedule.week_day}`);
  }

  // Extract active weeks from boolean flags
  const activeWeeks: number[] = [];
  if (schedule.week1) activeWeeks.push(1);
  if (schedule.week2) activeWeeks.push(2);
  if (schedule.week3) activeWeeks.push(3);
  if (schedule.week4) activeWeeks.push(4);
  if (schedule.week5) activeWeeks.push(5);

  if (activeWeeks.length === 0) {
    throw new Error('Schedule has no active weeks configured.');
  }

  // Search up to 13 months ahead
  for (let monthOffset = 0; monthOffset < 13; monthOffset++) {
    const monthIndex = reference.getMonth() + monthOffset;
    const year = reference.getFullYear() + Math.floor(monthIndex / 12);
    const month = (monthIndex % 12) + 1;

    for (const occurrence of activeWeeks.sort((a, b) => a - b)) {
      const day = nthWeekday(year, month, weekday, occurrence);
      if (day === null) continue;

      let startDt = new Date(year, month - 1, day, schedule.from_hour, 0);
      let endDt = new Date(year, month - 1, day, schedule.to_hour, 0);

      // Handle windows that cross midnight
      if (endDt <= startDt) {
        endDt = new Date(endDt.getTime() + 24 * 60 * 60 * 1000);
      }

      // Skip windows that already passed
      if (endDt <= reference) continue;

      if (startDt <= reference || startDt > reference) {
        return [startDt, endDt];
      }
    }
  }

  throw new Error('Unable to compute next sweep window within 12 months.');
}

/**
 * Format datetime to Pacific timezone ISO8601 string.
 * Example: "2025-01-15T08:00:00-08:00"
 */
export function formatPacificTime(date: Date): string {
  // Get the date parts in Pacific timezone
  const formatter = new Intl.DateTimeFormat('en-US', {
    timeZone: PACIFIC_TZ,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
    timeZoneName: 'short',
  });

  const parts = formatter.formatToParts(date);
  const getValue = (type: string) => parts.find(p => p.type === type)?.value || '';

  const year = getValue('year');
  const month = getValue('month');
  const day = getValue('day');
  const hour = getValue('hour');
  const minute = getValue('minute');
  const second = getValue('second');
  const timeZoneName = getValue('timeZoneName');

  // Determine offset from timezone name (PST = -08:00, PDT = -07:00)
  const offset = timeZoneName.includes('PDT') ? '-07:00' : '-08:00';

  return `${year}-${month}-${day}T${hour}:${minute}:${second}${offset}`;
}

/**
 * Convert military time (e.g., 900 = 9:00 AM, 1800 = 6:00 PM) to hours and minutes.
 *
 * @param militaryTime - Time in military format (e.g., 900, 1800)
 * @returns Tuple of [hours, minutes]
 */
function parseMilitaryTime(militaryTime: number): [number, number] {
  const hours = Math.floor(militaryTime / 100);
  const minutes = militaryTime % 100;
  return [hours, minutes];
}

/**
 * Get the JavaScript weekday (0=Sunday) from a Date in Pacific time.
 *
 * @param date - The date to check
 * @returns JavaScript weekday (0=Sunday, 6=Saturday)
 */
function getPacificWeekday(date: Date): number {
  const formatter = new Intl.DateTimeFormat('en-US', {
    timeZone: PACIFIC_TZ,
    weekday: 'short',
  });
  const weekdayStr = formatter.format(date).toLowerCase();

  const jsWeekdayMap: Record<string, number> = {
    'sun': 0,
    'mon': 1,
    'tue': 2,
    'wed': 3,
    'thu': 4,
    'fri': 5,
    'sat': 6,
  };
  return jsWeekdayMap[weekdayStr] ?? 0;
}

/**
 * Get date parts (year, month, day, hour, minute) in Pacific timezone.
 *
 * @param date - The date to extract parts from
 * @returns Object with year, month (1-12), day, hour, minute
 */
function getPacificDateParts(date: Date): { year: number; month: number; day: number; hour: number; minute: number } {
  const formatter = new Intl.DateTimeFormat('en-US', {
    timeZone: PACIFIC_TZ,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  });

  const parts = formatter.formatToParts(date);
  const getValue = (type: string) => parseInt(parts.find(p => p.type === type)?.value || '0', 10);

  return {
    year: getValue('year'),
    month: getValue('month'),
    day: getValue('day'),
    hour: getValue('hour'),
    minute: getValue('minute'),
  };
}

/**
 * Create a Date in Pacific timezone from date parts.
 *
 * @param year - Year
 * @param month - Month (1-12)
 * @param day - Day of month
 * @param hour - Hour (0-23)
 * @param minute - Minute (0-59)
 * @returns Date object
 */
function createPacificDate(year: number, month: number, day: number, hour: number, minute: number): Date {
  // Create a date string in ISO format with Pacific offset
  // We'll use a temporary date to determine if it's PST or PDT
  const tempDate = new Date(year, month - 1, day);
  const formatter = new Intl.DateTimeFormat('en-US', {
    timeZone: PACIFIC_TZ,
    timeZoneName: 'short',
  });
  const parts = formatter.formatToParts(tempDate);
  const tzName = parts.find(p => p.type === 'timeZoneName')?.value || 'PST';
  const offset = tzName.includes('PDT') ? '-07:00' : '-08:00';

  const isoString = `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}T${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}:00${offset}`;
  return new Date(isoString);
}

/**
 * Calculate the next move deadline for a time-limited parking regulation.
 *
 * Logic:
 * - If currently within a regulation window, deadline = now + hourLimit hours
 *   (but capped at window end time, rolling to next day if needed)
 * - If outside regulation window, deadline = next window start + hourLimit hours
 *
 * @param days - Days string ("M-F", "M-Sa", "M-Su")
 * @param hrsBegin - Start time in military format (e.g., 900 = 9:00 AM)
 * @param hrsEnd - End time in military format (e.g., 1800 = 6:00 PM)
 * @param hourLimit - Hour limit (e.g., 2 for 2-hour parking)
 * @param now - Reference time (defaults to current time)
 * @returns The deadline by which the car must be moved
 */
export function nextMoveDeadline(
  days: string,
  hrsBegin: number,
  hrsEnd: number,
  hourLimit: number,
  now?: Date,
): Date {
  const reference = now || new Date();
  const regulatedWeekdays = parseDaysString(days);

  const [beginHour, beginMinute] = parseMilitaryTime(hrsBegin);
  const [endHour, endMinute] = parseMilitaryTime(hrsEnd);

  // Convert our Weekday enum (0=Monday) to JS weekday (0=Sunday)
  // Our enum: Mon=0, Tue=1, Wed=2, Thu=3, Fri=4, Sat=5, Sun=6
  // JS:       Sun=0, Mon=1, Tue=2, Wed=3, Thu=4, Fri=5, Sat=6
  const jsRegulatedDays = new Set<number>();
  for (const weekday of regulatedWeekdays) {
    // Convert: ourWeekday + 1, with Sunday (6) becoming 0
    const jsDay = weekday === Weekday.SUN ? 0 : weekday + 1;
    jsRegulatedDays.add(jsDay);
  }

  /**
   * Find the next regulated day starting from a given date.
   */
  function findNextRegulatedDay(startDate: Date, includeToday: boolean): Date {
    const parts = getPacificDateParts(startDate);
    let { year, month, day } = parts;

    // Create a date for the start of the search day
    let searchDate = createPacificDate(year, month, day, beginHour, beginMinute);

    if (!includeToday) {
      // Move to next day
      searchDate = new Date(searchDate.getTime() + 24 * 60 * 60 * 1000);
    }

    // Search up to 7 days to find a regulated day
    for (let i = 0; i < 7; i++) {
      const jsWeekday = getPacificWeekday(searchDate);
      if (jsRegulatedDays.has(jsWeekday)) {
        // Return start of regulation window on this day
        const searchParts = getPacificDateParts(searchDate);
        return createPacificDate(searchParts.year, searchParts.month, searchParts.day, beginHour, beginMinute);
      }
      searchDate = new Date(searchDate.getTime() + 24 * 60 * 60 * 1000);
    }

    throw new Error('Could not find a regulated day within 7 days');
  }

  // Get current time in Pacific
  const nowParts = getPacificDateParts(reference);
  const currentMinutesSinceMidnight = nowParts.hour * 60 + nowParts.minute;
  const beginMinutesSinceMidnight = beginHour * 60 + beginMinute;
  const endMinutesSinceMidnight = endHour * 60 + endMinute;

  const jsCurrentWeekday = getPacificWeekday(reference);
  const isRegulatedDay = jsRegulatedDays.has(jsCurrentWeekday);

  // Check if currently within the regulation window
  const isWithinWindow =
    isRegulatedDay &&
    currentMinutesSinceMidnight >= beginMinutesSinceMidnight &&
    currentMinutesSinceMidnight < endMinutesSinceMidnight;

  if (isWithinWindow) {
    // Calculate tentative deadline: now + hourLimit hours
    const tentativeDeadline = new Date(reference.getTime() + hourLimit * 60 * 60 * 1000);
    const tentativeParts = getPacificDateParts(tentativeDeadline);
    const tentativeMinutes = tentativeParts.hour * 60 + tentativeParts.minute;

    // Check if tentative deadline is still within today's window
    // (same day and before end time)
    const isSameDay =
      tentativeParts.year === nowParts.year &&
      tentativeParts.month === nowParts.month &&
      tentativeParts.day === nowParts.day;

    if (isSameDay && tentativeMinutes <= endMinutesSinceMidnight) {
      return tentativeDeadline;
    }

    // Deadline exceeds today's window - find next regulated day
    const nextWindowStart = findNextRegulatedDay(reference, false);
    return new Date(nextWindowStart.getTime() + hourLimit * 60 * 60 * 1000);
  }

  // Not within window - check if before or after today's window
  if (isRegulatedDay && currentMinutesSinceMidnight < beginMinutesSinceMidnight) {
    // Before today's window starts - deadline is today's start + hourLimit
    const windowStart = createPacificDate(nowParts.year, nowParts.month, nowParts.day, beginHour, beginMinute);
    return new Date(windowStart.getTime() + hourLimit * 60 * 60 * 1000);
  }

  // After today's window (or non-regulated day) - find next regulated day
  const nextWindowStart = findNextRegulatedDay(reference, false);
  return new Date(nextWindowStart.getTime() + hourLimit * 60 * 60 * 1000);
}
